### ISSUE TITLE
Possible performance issue with dragvoer

### ISSUE THREAD

**Issue by danjford** [akserg/ng2-dnd#67]
* **I'm submitting a ...**
[x] bug report
[x] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**

A feature and possible also a bug.

* **What is the current behavior?**

In firefox (and probably other browsers), it seems that the dragover event is fired hundreds of times when an item is dragged over an element. This leads to a drop in FPS and it seems to be made worse depending on how many elements have had draggable / droppable applied.

The performance hit seems to be worse on firefox than e.g. chrome.

* **If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem** via https://plnkr.co or similar.

Taken using Firefox performance analyser:
![image](https://cloud.githubusercontent.com/assets/7963046/18873322/b896b5c8-84b6-11e6-9b85-9ee0a36b1eab.png)

Most of these events are dragover.

* **What is the expected behavior?**

Dragover not to fire so many times.

* **What is the motivation / use case for changing the behavior?**

Improved performance, keep the FPS from dropping so low.

* **Please tell us about your environment:**

- Browser: [all | Firefox 49.0.1 ]

* **Other information** (e.g. detailed explanation, stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, gitter, etc)

Suggestion how to fix: 

At the moment when `onDragOverCallback` it will only fire if `this._dragDropService.isDragged` is true. Then a class is added `this._elem.classList.add(this._config.onDragOverClass);`. Maybe a further check should be done so that the code is only fired if this class has not already been added to the target element e.g.:

```
    _onDragOverCallback (event: MouseEvent) {
        if (this._dragDropService.isDragged && !this._elem.classList.contains(this._config.onDragOverClass);) {
            this._elem.classList.add(this._config.onDragOverClass);
            this.onDragOver.emit({dragData: this._dragDropService.dragData, mouseEvent: event});
        }
    };
```

**Comment by akserg** [comment]
Hi @danjford 
Thank you for your investigation. I will have a look.

**Comment by joshvanal** [comment]
Hello!
So currently I am running into a performance issue and this looks related.  I have dnd-sortable being applied to a complex object because of the nature of this application. With 2 elements, if I start dragging an element between the 2 spots quickly, i noticed a ton of work was being on during the dragover event that would not let the page repaint quickly. I noticed a huge performance hit when I have 88 elements.

**Comment by joshvanal** [comment]
Any update on this @akserg

**Comment by akserg** [comment]
Hi @joshvanal 

I have the similar issue inside of my project as well. I have a fix and still test it until rest of the week.

**Comment by joshvanal** [comment]
Awesome @akserg. If there is anything I can do to help please let me know!

**Comment by alex88** [comment]
Same here, dragover event called so many times that makes the application freeze and takes like 3-4 seconds to show the new item position (dragging a div column in a list of 5 elements). Have you set a debounce as a fix?

**Comment by alan345** [comment]
Hi same issue here:

When I Drag in plunkr (http://embed.plnkr.co/JbG8Si), it works fine:

![image](https://user-images.githubusercontent.com/15246526/27771084-bbd0d568-5efc-11e7-82bd-f6b6adccc4f1.png)


But on my app:
![image](https://user-images.githubusercontent.com/15246526/27771070-805c9148-5efc-11e7-84a7-f894ec4debcc.png)
![image](https://user-images.githubusercontent.com/15246526/27771087-e0a09694-5efc-11e7-8fc3-ebc3bb089e99.png)

Memory get saturated.


I am using angular4 and bootstrap4
Thanks

**Comment by alex88** [comment]
@alan345 I have, and I have a block in the page that uses the auth service to check if the user is logged in, in my header it shows if the current user is logged in.
I had the same issue because I was checking with angular-jwt if the token was expired and it was resulting in some computation, I've fixed it temporarily by checking only page load and then just return the isLogged stored boolean. I think it would be better if I used an Observable using the async pipe.

However, seems the issue is that angular re-renders the whole page

**Comment by alexsalmon** [comment]
How's it going @akserg?

**Comment by CLEMARCx** [comment]
If someone needs a quick workaround:
Change the change detection to OnPush and trigger it manually.

`import { 
  ...
  ChangeDetectionStrategy
} from '@angular/core';

@Component({
   ...
  changeDetection: ChangeDetectionStrategy.OnPush, // set to OnPush
  ...
})
export class BoxComponent {
   ...
}`

**Comment by KDingeling** [comment]
Would be great to hear some news on this, currently I'm running into performance issues while dragging elements around, when there are ~8-10 elements in my sortable container. I see in the elements view of chrome that all draggable + droppable elements get updated permantly (divs are flashing in the DOM) while dragging an element anywhere on the screen. Will try to set the change detection to onPush for now.

**Comment by AnthonySaldana** [comment]
So i'm dealing with a similar issue where the dragging is just so laggy when theres a lot going on the page.
I have a few components that were repeated in a loop and I can drag it across those components but it's so laggy

The change detector seems to be the cause of the lag.
The workaround i did for this was to detach the change detector on drag start and to reattach it on drop. You could also set the change detection strategy to on push but this way you get to keep the change detector strategy and just turn it off while dragging.

`
constructor( private cd: ChangeDetectorRef) { }

onDragStart(){
    this.cd.detach();
}
onDrop(){
    this.cd.reattach();
}
`

**Comment by attodorov** [comment]
I am experiencing exactly the same issue, there's a severe performance hit, dropping is sometimes not even possible (takes more than 20-30 seconds to complete). It also depends how many child elements the draggable has. 

Is there an update on this issue? 

Thanks

**Comment by IdanCo** [comment]
Excellent Solutions!!

@CLEMARCx [solution](https://github.com/akserg/ng2-dnd/issues/67#issuecomment-326541270) is great, just don't forget to add `this.cd.markForCheck();` where needed (as explained [here](https://blog.thoughtram.io/angular/2016/02/22/angular-2-change-detection-explained.html))

@AnthonySaldana solution works great as well and much more targeted. Notice though that `onDragStart` and `onDragEnd` aren't documented (#195 ) but you can see them [here](https://github.com/akserg/ng2-dnd/blob/master/src/draggable.component.ts#L22). I also had to use [`onDragSuccess `](https://github.com/akserg/ng2-dnd/blob/master/src/draggable.component.ts#L34) for cases the drop event was inside a droppable area.

Both solutions add a little code ugliness, but it does the trick until @akserg will be more available (great project BTW)

**Comment by basileos** [comment]
i faced the same issue, because of having function's call inside draggable element, this function returns text or formatted date due to conditions, so it was called thousands times during dragging, i fixed this by placing already prepared data into the template.

**Comment by shashidhargm** [comment]
@KDingeling 
Hi, Even I'm facing the issue in my angular app. Please provide the solution if you have solved your issue.  
I am using Angular 4.4.3 and Primeng 4.2.1.

**Comment by KDingeling** [comment]
@shashidhargm 
Hey, I will take a look next week when I'm back at work. Can't remember it in detail.

**Comment by shashidhargm** [comment]
@KDingeling 
Thanks :)

**Comment by jmesa-sistel** [comment]
I had similar issue my solution is fork this lib and do the next modification in AbstractComponent

Change
```
this.elem.ondragover = (event: DragEvent) => {
     this._onDragOver(event);
     if (event.dataTransfer) {
         event.dataTransfer.dropEffect = this.config.dropEffect.name;
     }
     return false;
};
```
by
```
        this.ngZone.runOutsideAngular(() => {
            this.elem.ondragover = (event: DragEvent) => {
                this._onDragOver(event);
                if (event.dataTransfer) {
                    event.dataTransfer.dropEffect = this.config.dropEffect.name;
                }
                return false;
            };
        });
```
This change implies change the constructors to inject ngZone
I am not using dnd as a library I have copied all files in my app to do the tests.
Also I made my own changes to improve performance a little and clean code. But the main change to improve performance is this.
Also as @AnthonySaldana suggested I detach and reattach cdr in onDragStart / onDragEnd in the library, so I do not need to do it in all components that use dnd but I reduce detectChanges timeout to 20ms to get about 50fps

**Comment by joedjc** [comment]
@jmesa-sistel @akserg I also found running the drag events outside the Angular zone has significant performance benefits. There are still issues with lots of elements but this is much better. Created a fork here: https://github.com/joedjc/ng2-dnd

**Comment by jmesa-sistel** [comment]
@joedjc 
Remeber to execute the events inside angular
For example:
```
_onDragStartCallback(event: MouseEvent) {
        this.dragDropService.isDragged = true;
        this.dragDropService.dragData = this.dragData;
        this.dragDropService.onDragSuccessCallback = this.dragSuccess;
        setTimeout(() => {
            this.elem.classList.add(this.config.onDragStartClass);
        }, 0);
        this.ngZone.run(() => {
            this.dragStart.emit({ dragData: this.dragData, mouseEvent: event });
        });
    }
```
The seTtimeout fix another bug when you try to drag an element and you click near the border
I have rewrite the library to fix 4 or 5 bugs and I have clean the code and now the performance is not a problem. I will try to publish soon, when I have time, too busy at work. Too many changes to make a PR.

**Comment by apreeti** [comment]
> @jmesa-sistel @akserg I also found running the drag events outside the Angular zone has significant performance benefits. There are still issues with lots of elements but this is much better. Created a PR here: #269

It really is giving lots of performance benefits!! Thank you :)
